### PR TITLE
Patch fix

### PR DESCRIPTION
**Pull request by voellm**
This fix optimizes the driver by removing unneeded work and fixes a bug in update which was clobbering a document rather than updating it.

### CODE REVIEW

**Comment by stfeng2** [comment]
voellm@ - can you resolve the conflicts and push a new commit with the conflict resolved to this PR?

File: azurecosmos/src/main/java/com/yahoo/ycsb/db/AzureCosmosClient.java [hunk 1]
```diff
@@ -298,20 +284,41 @@ public Status scan(String table, String startkey, int recordcount, Set<String> f
 
   @Override
   public Status update(String table, String key, Map<String, ByteIterator> values) {
+    // Azure Cosmos does not have patch support.  Until then we need to read
+    // the document, update in place, and then write back.
+    // This could actually me made more efficient by using a stored procedure
```
Comment (stfeng2) [inline azurecosmos/src/main/java/com/yahoo/ycsb/db/AzureCosmosClient.java]: "could actually be made" (s/me/be)

File: azurecosmos/src/main/java/com/yahoo/ycsb/db/AzureCosmosClient.java [hunk 2]
```diff
@@ -298,20 +284,41 @@ public Status scan(String table, String startkey, int recordcount, Set<String> f
...
+
     String documentLink = getDocumentLink(this.databaseName, table, key);
-    Document document = getDocumentDefinition(key, values);
+    ResourceResponse<Document> updatedResource = null;
+    ResourceResponse<Document> readResouce = null;
+    RequestOptions reqOptions = null;
+    Document document = null;
+
+    try {
+      reqOptions = getRequestOptions(key);
```
Comment (stfeng2) [inline azurecosmos/src/main/java/com/yahoo/ycsb/db/AzureCosmosClient.java]: can we just embed the getRequestOptions() call into the readDocument() call? So the local variable reqOptions won't need to be introduced here.

File: cassandra/src/main/java/com/yahoo/ycsb/db/CassandraCQLClient.java [hunk 3]
```diff
@@ -171,9 +174,17 @@ public void init() throws DBException {
             getProperties().getProperty(WRITE_CONSISTENCY_LEVEL_PROPERTY,
                 WRITE_CONSISTENCY_LEVEL_PROPERTY_DEFAULT));
 
+        Boolean useSSL = Boolean.parseBoolean(getProperties().getProperty(USE_SSL_CONNECTION,
+            DEFAULT_USE_SSL_CONNECTION));
+
         if ((username != null) && !username.isEmpty()) {
-          cluster = Cluster.builder().withCredentials(username, password)
-              .withPort(Integer.valueOf(port)).addContactPoints(hosts).build();
+          if (useSSL) {
```
Comment (stfeng2) [inline cassandra/src/main/java/com/yahoo/ycsb/db/CassandraCQLClient.java]: Take this part away, it's conflicting with a previous PR...

File: azurecosmos/src/main/java/com/yahoo/ycsb/db/AzureCosmosClient.java (hunk 2 above)
Comment (voellm) [reply azurecosmos/src/main/java/com/yahoo/ycsb/db/AzureCosmosClient.java]: Its actually used a couple of times in this function so I avoid creating multiple objects.

File: cassandra/src/main/java/com/yahoo/ycsb/db/CassandraCQLClient.java (hunk 3 above)
Comment (voellm) [reply cassandra/src/main/java/com/yahoo/ycsb/db/CassandraCQLClient.java]: Fixed.  The perils of multiple changes in flight.

File: azurecosmos/src/main/java/com/yahoo/ycsb/db/AzureCosmosClient.java (hunk 1 above)
Comment (voellm) [reply azurecosmos/src/main/java/com/yahoo/ycsb/db/AzureCosmosClient.java]: Fixed.  Thanks.

**Comment by voellm** [comment]
Abandoning this to redo it.  Too many non related changes got picked up.

**Pull request closed without merging**